docs: Introduce language implementation handbook (#1428)
aws/jsii: docs/handbooks/language-implementation.md

# Language Implementation Handbook

> 

Reviewer: Add a sentence or 2 on what this is and who it's intended for.

## Foreword

Implementing a new language in *jsii* is not just a matter of implementing code

> generation: mapping the *[jsii type system]* to a new programming language takes

Reviewer: ```suggestion
generation. Mapping the *[jsii type system]* to a new programming language takes
```

> some amounts of API design, finding ways to represent APIs in the most idiomatic

Reviewer: Either prephase what API is or use a different/more appropriate term here. Maybe 'language interface design' ?

possible way. This is a craft that often requires trial and error, and the best
(if not only) way to validate a proposal is to put it in front of users and
seek feedback. As a consequence, this endeavor should be expected to span
months, not weeks.

## Scoping & Planning

The first step of most successful projects is to start by scoping work out and
establishing a baseline plan to execute on. For contributors not yet familiar
with *jsii*, the [specification] document is a great place to start. In
particular, the [New Language Intake] document provides a high-level view of the
recommended process for implementing new language support.

> 

Reviewer 2: I would add a table here that can be used as a TOC and summarizes the recommended activities and some rough estimation of how long they take in developer days/weeks

### Language Proposition RFC

The recommended way to formalize the initial plan is to write it into an RFC
hosted in the [CDK RFC repository]. Enough time has to be spent considering the
requirements in order to get the work scoped and planned well, ensuring smooth
execution.

An additional benefit of following the RFC process is that it makes it easier
to track learnings accumulated through the implementation process, as those will
be tracked as comments or iterations on the RFC document.

It is possible (and sometimes desirable) to start prototyping code-generation
for the new language, as this can hlighlight implementation challenges that need
to be discussed in the RFC document. In any case, examples of the API signatures
that are expected to be rendered allow early feedback to be provided by possible
future users, and still helps identify challenges.

The following questions should be answered as early as possible in the process,
to avoid surprises later on that result in significant re-engineering effort:

> * What does the generated APIs look like, for the typical API idioms?

Reviewer: ```suggestion
* What do the generated APIs look like, for the typical API idioms?
```

>   - *Classes* (constructors, properties, methods, inheritance strategy, abstract

Reviewer 2: I wonder if this should be an exhaustive list that can serve as a checklist/template for the RFC

Author: I surely intend it to become that, but wanted to avoid sounding too prescriptive before I have at least 1 RFC to look up to :)

    members, ...)

>     + Construction experience when last argument of constructor is a struct

Reviewer: This sentence is suddenly dropping too low into the implementation without context.

Or build a bit of context. "In the CDK, with the biggest adoption of JSII, follows a pattern where every the constructor of most classes has a struct as its last parameter (e.g., ...)...."

  - *Enums*
  - *Interfaces* and *Structs* (properties, methods, inheritance strategy,
    implementation, ...). In particular, how are new optional properties handled
    (those are not considered breaking change within the [jsii type system]).
  - *Structs* (properties, inheritance strategy, implementation, ...)
* What information is needed in order for the code-generator to produce

>   artifacts? What should the configuration block look like?

Reviewer: What's a configuration block? (half rhetorical to point out that as a reader I'm not yet familiar with this. the other half because I genuinely don't know 😉)

Author: Added a pitch to this earlier in the document.

* What is the standard way to publish packages for the new language?

>   - Is there any requirements (code signature, special metadata, ...) that need

Reviewer: ```suggestion
  - Are there any requirements (code signature, special metadata, ...) that need
```

    to be implemented in order to publish valid packages?

>   - How are dependencies modeled? If semantic versioning  is not the norm, what

Reviewer: Link to semantic versioning

    is the strategy to correctly represent semantic version ranges?

> * What are the toolchain and platform requirements?

Reviewer: Couple of examples (in parenthesis) from our currently implemented langs would be useful.

## Code Generation

> Before investing too much work towards the *host library* implementation, it is

Reviewer 2: Perhaps the overview section should provide a high level view of which components are going to be needed: 

- code generator
- runtime library
- compliance tests
- packager
- rossetta
- publisher (see the npm module jsii-release)

> recommended to write a first version of the code generation for the new language.

Reviewer: I would rephrase to "First, implement the first version of the code generation for the new language before getting too far into the *host library* implementation."

Also explain with a sentence on why this is recommended.

This work happens within the [`jsii-pacmak`] package.

Initially, focus on the API signatures: for example, generate code that throws
a *not implemented* exception when called. Start by making sure a decent API is
generated from the [`jsii-calc`] package and its dependencies, and use those to

> implement the tests form the [Standard Compliance Suite].

Reviewer: ```suggestion
Focus, initially, on generating correct API signatures, before getting into their 
implementation. The first version may even simply throw a *not implemented*
exception when called. 

The [`jsii-calc`] package can be used a sample consuming library which uses
JSII to generate code in all target languages. Start by making sure a decent API is
generated from this package and its dependencies, and use those to
implement the tests from the [Standard Compliance Suite].
```

## Host library

Now that we are generating "empty shell" APIs that represent the necessary
entities to back the [Standard Compliance Suite] tests, start implementing the
*host library* and update the code generator to make the tests pass until all
of them pass. It's possible to start publishing artifacts before all the tests

> pass. As soon as some basic features are working, work on [Building and

Reviewer: ```suggestion
*host library* and update the code generator until all of the tests pass. 
It's possible to publish artifacts even when tests in the suite are failing.
As soon as some basic features are working, work on [Building and
```

Packaging](#building-and-packaging) can start, so early feedback can be
gathered.

> :construction: A standard architecture for the *host library* has not been
> documented yet. Upcoming language implementations should contribute to this
> process by documenting a general architecture that should be implementable
> in any programming languages (and thus, asbtracting away language
> specificities).

> ## Building and Packaging

Reviewer: Should we also mention versioning these packages as 'experiemental'?

The necessary toolchains should be added to he [`jsii/superchain`] Docker image,
so that the [`jsii-pacmak`] generation can be changed to support building ready
to publish artifacts instead of just code.

> Additionally, new constructs have to be added to [`aws-delivlib`] in order to

Reviewer: ```suggestion
Additionally, [`aws-delivlib`] needs to be augumented to support publishing artifacts to the language's package repository.
```

Reviewer 2: I’ve extracted this to jsii-release (still on eladb/jsii-release)

support building and publishing of those new artifacts.

> ## Documentation

Reviewer 2: What about the CDK reference documentation generator? I know it’s not in the jsii repo but it definitely should be!

Author: The **CDK** reference documentation generator is out-of-scope. A generic *jsii* documentation site generator would be in scope, but such a thing does not exist and I cannot reasonably set expectations that one will implement all of this from scratch as part of a new language.

Before releasing the new language support to *Developer Preview*, basic
documentation needs to be produced to explain how to configure a *jsii* project
to support the new language, and any peculiarities in working with libraries
generated by [`jsii-pacmak`] for this language.

## Developer Preview

Once the full [Standard Compliance Suite] passes (possibly with the exception of
certain fringy features), and the documentation covering all aspects of using
the language bindings have been produced, the new language can be released to
*Developer Preview*.

It is recommended that new languages stay in *Developer Preview* for a minimum

> of 4 weeks, or until they have received sufficient usage to have built

Reviewer: 4 weeks feels very low, but ok. Should be at least double that.

Author: Replaced the `or until` with `ideally until`.
